### PR TITLE
Feature - Request Tracking

### PR DESCRIPTION
**Pull request by cnoon**
This PR refactors the fragile activity counting to a more robust request id tracking system.

### Issue Link :link:
#38 

### Goals :soccer:
The goals of this PR are to:
- Fix issue #38
- Create a better public API that does not rely on a single counter

### Implementation Details :construction:
The original implementation relied on an activity counter that would run the activity indicator when the count was greater than zero. This proved to be too fragile with mismatches between resume, suspend, and complete calls as well as the public APIs for controlling the count.

This PR switches away from using a single counter to instead tracking active request through a set of `String` requestIDs. I chose to go with a `Set<String>` instead of `Set<UUID>` because other clients generating requests may not be generating UUIDs alongside them. I thought it would make it easier for them to adopt if we used `String` identifiers instead.

### Testing Details :mag:
I've updated the test suite to test all the new edge cases with the new APIs.

### CODE REVIEW

File: Source/NetworkActivityIndicatorManager.swift [hunk 1]
```diff
@@ -124,31 +126,29 @@ public class NetworkActivityIndicatorManager {
...
+    /// Adds the requestID as an active request driving the activity indicator.
     ///
-    /// If this number was zero before incrementing, the network activity indicator will start spinning after
-    /// the `startDelay`.
+    /// This method results in a no-op if the request is already being tracked.
     ///
-    /// Generally, this method should not need to be used directly.
-    public func incrementActivityCount() {
+    /// - Parameter requestID: The request identifier.
+    public func requestDidStart(requestID: String) {
```
Comment (cnoon) [inline Source/NetworkActivityIndicatorManager.swift]: I'd love to get your thoughts on this API @jshier. I went through a lot of iterations on the name of this API, and landed on `requestDidStart` and `requestDidStop`.

File: Source/NetworkActivityIndicatorManager.swift [hunk 2]
```diff
@@ -203,12 +203,14 @@ public class NetworkActivityIndicatorManager {
 
     // MARK: - Private - Notifications
 
-    @objc private func networkRequestDidStart() {
-        incrementActivityCount()
+    @objc private func networkRequestDidStart(notification: Notification) {
+        guard let request = notification.request else { return }
```
Comment (cnoon) [inline Source/NetworkActivityIndicatorManager.swift]: Handy extension!

File: Source/NetworkActivityIndicatorManager.swift (hunk 1 above)
Comment (jshier) [reply Source/NetworkActivityIndicatorManager.swift]: Seems like the only other option would be `didStartRequest(id: String)`. Either way, we may want to make the parameter name `id`, as `requestID` seems a bit redundant.

File: Source/NetworkActivityIndicatorManager.swift [hunk 3]
```diff
@@ -87,6 +87,8 @@ public class NetworkActivityIndicatorManager {
 
     private var activityIndicatorState: ActivityIndicatorState = .notActive {
         didSet {
+            lock.lock() ; defer { lock.unlock() }
```
[the remark this answers is not in the document]
Comment (jshier) [reply Source/NetworkActivityIndicatorManager.swift]: Is there a path to the `didSet` that isn't through a method that's already locked? They're run synchronously, so they'll be run before the `defer` unlocks I think.

[the remark this answers is not in the document]
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Good catch. I couldn't remember if `didSet` was run synchronously or not. There's no path that isn't currently locked, so this change technically isn't necessary then. The bigger question is whether we'd want to protect against a future change that mutated the state that wasn't locked. I could go either way on that one since I updated the lock to be recursive, but I think I'll remove it for now since it's not technically needed.

File: Source/NetworkActivityIndicatorManager.swift (hunk 1 above)
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Yep. We have a convention at Nike where we would use `requestDidStart(withID requestID: String)` where we drop the `request` portion of the external API and supplement with `with` to make it read better. Then we keep the `request` on the internal parameter. Is that a convention that would be a better alternative than just `id`? We try to avoid `id` so that we always know exactly what `id` actually represents. In this case, our intention is that the `id` represents the `requestID`, but to your point, that can certainly be inferred given the rest of the API signature.

Anyways, I'm happy with any of the following:

- `requestDidStart(requestID: String) // option 1`
- `requestDidStart(id: String) // option 2`
- `requestDidStart(withID requestID: String) // option 3`

Do you have a strong preference on any?

File: Source/NetworkActivityIndicatorManager.swift [hunk 4]
```diff
@@ -177,21 +177,21 @@ public class NetworkActivityIndicatorManager {
...
-            selector: #selector(NetworkActivityIndicatorManager.networkRequestDidStart),
+            selector: #selector(NetworkActivityIndicatorManager.networkRequestDidStart(notification:)),
             name: Request.didResume,
             object: nil
         )
 
         notificationCenter.addObserver(
             self,
-            selector: #selector(NetworkActivityIndicatorManager.networkRequestDidComplete),
+            selector: #selector(NetworkActivityIndicatorManager.networkRequestDidStop(notification:)),
```
[the remark this answers is not in the document]
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Meaning I could drop the `NetworkActivityIndicatorManager` portion of the API? I'll dig into it.

[the remark this answers is not in the document]
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Removed the `(notification:)` portion of the selectors in cd21ed89.

File: Source/NetworkActivityIndicatorManager.swift (hunk 1 above)
Comment (jshier) [reply Source/NetworkActivityIndicatorManager.swift]: I guess it depends on whether you consider the `with` implicit in the fact that it's a parameter. I think 2 or 1 would work, depending on whether we need the disambiguation for `id`.

File: Source/NetworkActivityIndicatorManager.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Dropped this unnecessary lock in 0e657077.

File: Source/NetworkActivityIndicatorManager.swift (hunk 1 above)
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Refactored API to `requestDidStart(id: String)` in bf277d73.

**Comment by cnoon** [comment]
Couldn't actually drop the `NetworkActivityIndicatorManager` portion inside the `#selector`.

File: Source/NetworkActivityIndicatorManager.swift (hunk 1 above)
Comment (cnoon) [reply Source/NetworkActivityIndicatorManager.swift]: Decided to stick with `requestDidStart(requestID: String)` based on Slack conversation to avoid ambiguity and make it very clear that the `id` is meant to be the `requestID`.
